src/dmd/filecache.d

```diff
@@ -0,0 +1,137 @@
+/**
+ * Compiler implementation of the
+ * $(LINK2 http://www.dlang.org, D programming language).
+ *
+ * Copyright:   Copyright (C) 1999-2018 by The D Language Foundation, All Rights Reserved
+ * Authors:     $(LINK2 http://www.digitalmars.com, Walter Bright)
+ * License:     $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
+ * Source:      $(LINK2 https://github.com/dlang/dmd/blob/master/src/dmd/filecache.d, _errors.d)
+ * Documentation:  https://dlang.org/phobos/dmd_filecache.html
+ * Coverage:    https://codecov.io/gh/dlang/dmd/src/master/src/dmd/filecache.d
+ */
+
+module dmd.filecache;
+
+import dmd.root.stringtable;
+import dmd.root.array;
+import dmd.root.file;
+
+import core.stdc.stdio;
+
+// A simple wrapper around a string in D for extern(C++) compatibility
+extern(C++) struct Dstring
+{
+    char* ptr;
+    size_t length;
+    extern(D) string toString()
+    {
+        return cast(string) ptr[0 .. length];
+    }
+}
+
+/**
+A line-by-line representation of a $(REF File, dmd, root, file).
+*/
+class FileAndLines
+{
+    File* file;
+    Array!Dstring* lines;
+
+    /**
+    File to read and split into its lines.
+    */
+    this(const(char)* filename)
+    {
+        file = File.create(filename);
+        lines = new Array!Dstring;
+        readAndSplit();
+    }
+
+    // Read a file and split the file buffer linewise
+    private void readAndSplit()
+    {
+        file.read();
+        auto buf = file.buffer;
+        // slice into lines
+        while (*buf)
+        {
+            auto prevBuf = buf;
+            for (; *buf != '\n'; buf++)
```

Comment:
Do we need to care about Windows line endings?

Reply:
Yeah. Added a check for it.